Fixed #31124 -- Fixed setting of get_FOO_display() when overriding inherited choices.

carltongibson opened the pull request:
(To be squashed and tidied obviously)

Three commits: 

* Test for ticket-31124
* Initial fix, but...
* Further test, since it can come up again. 

@felixxm @claudep @sir-sigurd: What do you all think? 

I [initially suggested](https://github.com/django/django/commit/e3c035ef177d9ed40340b2d294aa55a7e5100529#diff-5b33a9c46f488003c1846ef677f861d3R820-R839) for ticket-30931 that we doc overriding `_get_FIELD_display()`:

```
.. admonition:: Overriding extra instance methods

    In order to customize ``get_FOO_display()``, or either of
    ``get_next_by_FOO()`` or ``get_previous_by_FOO()``, you need to override
    the helper methods ``_get_FIELD_display()``, or
    ``_get_next_or_previous_by_FIELD()``, respectively.

    For example, you might customize ``get_FOO_display()`` like so::

        from django.db import models


        class FooBar(models.Model):
            foo_bar = models.IntegerField(choices=[(1, 'foo'), (2, 'bar')])

            def _get_FIELD_display(self, field):
                if field.attname == 'foo_bar':
                    return 'custom display string'
                return super()._get_FIELD_display(field)
```

I suggested this saying, about the fix in `contribute_to_class()`, that "this way lies madness", i.e. we're in danger of adding ever more complicated logic into `contribute_to_class()`, when there's a clean way to point users to available. 

We decided to add the `hasattr` check, which is OK.

Now we can move to inspecting the `__dict__` directly (to avoid walking the bases). I guess that might be OK too. 

But we need to decide at what point we cut our loses and say that we don't support this. So...

Is this further fix OK? Do we need to support the last test case? If not do we doc it? OR do we back out and doc the non-`__new__()` approach? Or...? (My inclination is towards the simpler code, and so documenting overriding `_get_FIELD_display()`—for me, it's just trying to be too clever here—but happy to go with majority opinion.)

Thanks for the thoughts! 
(And happy new year! 🎊)

felixxm commented:
@carltongibson Thanks for this patch :+1: We're falling down the deep rabbit hole :rabbit:

Overriding `get_FOO_display()`, `get_next_by_FOO()`, or  `get_previous_by_FOO()` is confusing, and I think we shouldn't support it at all. If someone want to customize their behavior it is always possible with a custom method added in `contribute_to_class()`. So I would suggest to revert 2d38eb0ab9f78d68c083a5b78b1eca39027b279a and add a note/warning :fire:  that it's not supported :red_circle: .

IMO, adding [suggestion](https://github.com/django/django/commit/e3c035ef177d9ed40340b2d294aa55a7e5100529#diff-5b33a9c46f488003c1846ef677f861d3R820-R839) to override methods that are a part of private API shouldn't be documented.

carltongibson commented:
Hey @felixxm. I've rebased this, added a comment in the docs, and the code, suggesting `contribute_to_class()`. I think we can fix this and say no-more. What do you think?

felixxm commented on django/db/models/fields/__init__.py:
```diff
@@ -764,7 +764,11 @@ def contribute_to_class(self, cls, name, private_only=False):
+            # Do not override a get_FOO_display defined explicitly on this
+            # class, but do not walk the inheritance tree, to allow overriding
+            # inherited choices. For more complex inheritance structures users
+            # should override contribute_to_class().
```
Maybe:
```
            # Don't override a get_FOO_display() method defined explicitly on
            # this class, but don't check methods derived from inheritance, to
            # allow overriding inherited choices. For more complex inheritance
            # structures users should override contribute_to_class().
```

felixxm commented on tests/model_fields/tests.py:
```diff
@@ -178,6 +178,22 @@ def get_foo_bar_display(self):
+        for t in tests:
+            with self.subTest(test=t):
+                c = Child(foo=t)
+                self.assertEqual(c.get_foo_display(), 'Child %s' % t)
```
I think we can remove `('B', 'Base B')` and `('B', 'Child B')` because it tests the same case as `A`.

felixxm replied:
IMO using `subTest()` is not necessary here, e.g.
```python
self.assertEqual(Child(foo='A').get_foo_display(), 'Child A')
self.assertEqual(Child(foo='B').get_foo_display(), 'Child B')
```
is more readable.

felixxm commented on docs/ref/models/instances.txt:
```diff
@@ -834,6 +834,15 @@ Note that in the case of identical date values, these methods will use the
 
+.. admonition:: Overriding extra instance methods
+
+    In simple cases overriding or inheriting ``get_FOO_display()``,
```
`In simple cases` -> `For most of cases`

felixxm commented on docs/ref/models/instances.txt:
```diff
@@ -834,6 +834,15 @@ Note that in the case of identical date values, these methods will use the
+
+    In simple cases overriding or inheriting ``get_FOO_display()``,
+    ``get_next_by_FOO()``, and ``get_previous_by_FOO()` should work as
+    expected. Since they are added by the metaclass, however, it is not
```
`     expected, since they are added by the metaclass. However, it is not `

carltongibson replied:
This one reverses the intended meaning. 

I want to say that the reason we can't handle all cases is precisely because the methods are added by the meta class. (Not that they work because they're added...)

felixxm replied:
Ahh, OK.

felixxm commented on docs/releases/3.0.3.txt:
```diff
@@ -31,3 +31,7 @@ Bugfixes
+
+* Fixed a regression in Django 2.2.7 that caused
+  :meth:`~django.db.models.Model.get_FOO_display` to work incorrectly when
+  overriding inherited choices.
```
``... choices (:ticket:`31124`). ``
